snapshots/windows/common.go

```diff
@@ -0,0 +1,419 @@
+			// snapshot if this isn't the snapshot that just houses the final sandbox.vhd
+			// that will be mounted as the containers scratch. Currently the key for a snapshot
+			// where a layer will be extracted to will have the string `extract-` in it.
+			return nil
+		}
+
+		if len(newSnapshot.ParentIDs) == 0 {
+			// A parentless snapshot a new base layer. Valid base layers must have a "Files" folder.
+			// When committed, there'll be some post-processing to fill in the rest
+			// of the metadata.
+			filesDir := filepath.Join(snDir, "Files")
+			if err := os.MkdirAll(filesDir, 0700); err != nil {
+				return fmt.Errorf("creating Files dir: %w", err)
+			}
+			return nil
+		}
+
+		parentLayerPaths := w.parentIDsToParentPaths(newSnapshot.ParentIDs)
+		var snapshotInfo snapshots.Info
+		for _, o := range opts {
+			o(&snapshotInfo)
+		}
+
+		var sizeInBytes uint64
+		if sizeGBstr, ok := snapshotInfo.Labels[rootfsSizeInGBLabel]; ok {
+			log.G(ctx).Warnf("%q label is deprecated, please use %q instead.", rootfsSizeInGBLabel, rootfsSizeInBytesLabel)
+
+			sizeInGB, err := strconv.ParseUint(sizeGBstr, 10, 32)
+			if err != nil {
+				return fmt.Errorf("failed to parse label %q=%q: %w", rootfsSizeInGBLabel, sizeGBstr, err)
+			}
+			sizeInBytes = sizeInGB * 1024 * 1024 * 1024
+		}
+
+		// Prefer the newer label in bytes over the deprecated Windows specific GB variant.
+		if sizeBytesStr, ok := snapshotInfo.Labels[rootfsSizeInBytesLabel]; ok {
+			sizeInBytes, err = strconv.ParseUint(sizeBytesStr, 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to parse label %q=%q: %w", rootfsSizeInBytesLabel, sizeBytesStr, err)
+			}
+		}
+
+		var makeUVMScratch bool
+		if _, ok := snapshotInfo.Labels[uvmScratchLabel]; ok {
+			makeUVMScratch = true
+		}
+
+		// This has to be run first to avoid clashing with the containers sandbox.vhdx.
+		if makeUVMScratch {
+			if err = w.createUVMScratchLayer(ctx, snDir, parentLayerPaths); err != nil {
+				return fmt.Errorf("failed to make UVM's scratch layer: %w", err)
+			}
+		}
+		if err = w.createScratchLayer(ctx, snDir, parentLayerPaths, sizeInBytes); err != nil {
+			return fmt.Errorf("failed to create scratch layer: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return w.mounts(newSnapshot, key), nil
+}
+
+// This is essentially a recreation of what HCS' CreateSandboxLayer does with some extra bells and
+// whistles like expanding the volume if a size is specified.
+func (w *windowsBaseSnapshotter) createScratchLayer(ctx context.Context, snDir string, parentLayers []string, sizeInBytes uint64) error {
+	parentLen := len(parentLayers)
+	if parentLen == 0 {
+		return errors.New("no parent layers present")
+	}
+
+	baseLayer := parentLayers[parentLen-1]
+	templateDiffDisk := filepath.Join(baseLayer, "blank.vhdx")
+	dest := filepath.Join(snDir, "sandbox.vhdx")
+	if err := copyScratchDisk(templateDiffDisk, dest); err != nil {
+		return err
+	}
```

Comment:
For cimfs, shouldn't we have a single template vhd (or one for each size) for all scratch layers, rather than storing in the base layer?

Reply:
From reading  https://github.com/microsoft/hcsshim/blob/main/internal/wclayer/cim/process.go#L23, it appears that the blank-base is in the base layer still, rather than being in the shared cim-fs directory, and the differencing VHD is relative to that base file, so I assume that template is different for each base layer. I'm not familiar with how differencing VHDs work here though, so maybe that's something that can be optimised, but it would need to be done in hcsshim first.

From memory, one of the advantages of having per-base template files like this is that we don't need any locking for the creation of those template files beyond existing per-snapshot locking, i.e. you can't operate a snapshot in the middle of an import. I suspect cimfs already breaks this assumption through management of the shared cimfs folder, so if we can share blank-base and blank vhdx files across bases, then cimfs supports it more naturally than the wcifs snapshotter, with the only cost that (partially) reforks the `create\*ScratchLayer` methods.

That (and the function comment...) suggests that this function is an abstraction break, it (and `createUVMScratchLayer`) relies on details that ought to be encapsulated by the hcsshim code. Alternatively, things like "blank.vhdx" and "SystemTemplate.vhdx" should be a public symbol (currently privately `<hcsshim>/internal/wclayer.ContainerScratchVhd` and friends but could be trivially exported) rather than hard-coded here to match a constant in hcsshim.

Reply:
@kevpar  Yeah, that's what I am working on right now but it's an optimization and so not strictly needed for 2.0. I am planning to get it merged with some other changes (like the layer management refactor) that we had in mind.

@TBBle what locks are you referring to here? Could you point me to the code?

Reply:
On my phone right now, but my recollection is that there's a boltdb transaction open or lock held that prevents simultaneous mutating operations. We had a TODO in one of the snapshotter PRs because the Windows snapshotter holds this lock in Commit unusually long compared to other snapshotters, as Windows restreams the layer, and for every other snapshotter Commit is only a metadata update.

Reply:
https://github.com/containerd/containerd/pull/4415#discussion_r612344502 is the boltdb locking discussion I was referring to above, which lead to the first time I lodged #5384.

The lock in question is all the calls to `WithTransaction` in https://github.com/containerd/containerd/blob/main/snapshots/windows/windows.go; it's a snapshotter-wide RWLock. [We don't actually need to use that MetaStore implementation](https://github.com/containerd/containerd/blob/main/snapshots/storage/metastore.go#L61-L65), if we have different locking requirements. (I think I saw that overlay snapshotter uses its own metadata storage implementation, but it was only in passing)

Reply:
I see. I think there are 2 different cases here. One is the scratch VHD creation that we do while holding the snapshotter lock. I think copying the VHD should be quick since for CimFS the VHD is all empty (no reparse points etc. like WCIFS). So even if we are holding the lock during that copy operation we should be good.

Second scenario is when we are exporting a scratch layer. The export operation of a scratch layer is time consuming and we don't want to keep holding the lock for that long. For CimFS, I believe we can convert the scratch layer into a CIM, in a temporary directory without holding any locks and then only take the lock when we want to update the metadata. However, none of this is implemented yet.

Reply:
Yeah, the main issue with releasing the lock for long-running operations (i.e. Commit) is that we need to ensure that a second call to Commit or a call to Mount doesn't happen (or fails when called) during that operation. We can't simply update the metadata early, because then someone could try to use the layer as a parent while it's still processing, and containerd doesn't have a layer state value for "Busy" or the like. This is all fairly solvable; at the time I was looking into it, we hadn't landed the mount support so it was waiting behind that to have integration test coverage before trying to optimise the internals. (Actually I wonder now what happens if you try to Commit or Mount a being-Committed layer. I guess you block at `WithTransaction` call, but when the in-progress Commit completes, does it just continue into the code and possibly corrupt stuff? Ew... I don't have the code in front of me, I hope we are doing sanity/safety checks inside the transaction.)